src/start/RealmInputScreen.js

```diff
@@ -33,88 +29,130 @@ const urlFromInputValue = (realmInputValue: string): URL | void => {
+
+  const [progress, setProgress] = React.useState(false);
+  const [realmInputValue, setRealmInputValue] = React.useState('');
+  const [error, setError] = React.useState(null);
 
-  tryRealm: () => Promise<void> = async () => {
-    const { realmInputValue } = this.state;
+  const textInputRef = React.useRef<React$ElementRef<typeof TextInput> | null>(null);
 
+  // When the route is focused in the navigation, focus the input.
+  // Otherwise, if you go back to this screen from the auth screen, the
+  // input won't be focused.
+  useFocusEffect(
+    useCallback(() => {
+      if (textInputRef.current) {
+        // Sometimes the effect of this `.focus()` is immediately undone
+        // (the keyboard is closed) by a Keyboard.dismiss() from React
+        // Navigation's internals. Seems like a complex bug, but the symptom
+        // isn't terrible, it just means that on back-navigating to this
+        // screen, sometimes the keyboard flicks open then closed, instead
+        // of just opening. Shrug. See
+        //   https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/realm-input/near/1346690
+        textInputRef.current.focus();
+      }
+    }, []),
+  );
+
+  const tryRealm = React.useCallback(async () => {
     const parsedRealm = urlFromInputValue(realmInputValue);
     if (!parsedRealm) {
-      this.setState({ error: 'Please enter a valid URL' });
+      setError('Please enter a valid URL');
       return;
     }
     if (parsedRealm.username !== '') {
-      this.setState({ error: 'Please enter the server URL, not your email' });
+      setError('Please enter the server URL, not your email');
       return;
     }
 
-    this.setState({
-      progress: true,
-      error: null,
-    });
+    setProgress(true);
+    setError(null);
     try {
       const serverSettings: ApiResponseServerSettings = await api.getServerSettings(parsedRealm);
-      this.props.navigation.push('auth', { serverSettings });
+      navigation.push('auth', { serverSettings });
       Keyboard.dismiss();
     } catch (errorIllTyped) {
       const err: mixed = errorIllTyped; // https://github.com/facebook/flow/issues/2470
-      this.setState({ error: 'Cannot connect to server' });
+      setError('Cannot connect to server');
       /* eslint-disable no-console */
       console.warn('RealmInputScreen: failed to connect to server:', err);
       // $FlowFixMe[incompatible-cast]: assuming caught exception was Error
       console.warn((err: Error).stack);
     } finally {
-      this.setState({ progress: false });
+      setProgress(false);
     }
-  };
-
-  handleRealmChange: string => void = value => this.setState({ realmInputValue: value });
+  }, [navigation, realmInputValue]);
 
-  render(): Node {
-    const { navigation } = this.props;
-    const { progress, error, realmInputValue } = this.state;
+  const handleRealmChange = useCallback(value => {
+    setRealmInputValue(value);
+  }, []);
```

Comment:
Can simplify this (in a followup commit) to just `setRealmInputValue` -- just refer to that directly instead of adding the name `handleRealmChange`.

Reply:
Ah yeah; I did that in an earlier version of this work, and meant to bring it here but forgot: https://github.com/zulip/zulip-mobile/pull/5328/commits/9afdc4cc71b4104ef4b9bbcbe4fbace3eebafa08